### PR TITLE
[BE] Clean up documentation for nn.Transformer related classes

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114789

### CODE REVIEW

File: torch/nn/modules/transformer.py [hunk 1]
```diff
@@ -126,14 +127,14 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
         r"""Take in and process masked source/target sequences.
 
         Args:
-            src: the sequence to the encoder (required).
-            tgt: the sequence to the decoder (required).
-            src_mask: the additive mask for the src sequence (optional).
-            tgt_mask: the additive mask for the tgt sequence (optional).
-            memory_mask: the additive mask for the encoder output (optional).
-            src_key_padding_mask: the Tensor mask for src keys per batch (optional).
```
Comment (janeyx99) [inline torch/nn/modules/transformer.py:134]: how come we remove the optional keyword in the new version?

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -158,34 +159,36 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
...
 
-            Note: [src/tgt/memory]_mask ensures that position i is allowed to attend the unmasked
-            positions. If a BoolTensor is provided, positions with ``True``
-            are not allowed to attend while ``False`` values will be unchanged. If a FloatTensor
-            is provided, it will be added to the attention weight.
-            [src/tgt/memory]_key_padding_mask provides specified elements in the key to be ignored by
-            the attention. If a BoolTensor is provided, the positions with the
-            value of ``True`` will be ignored while the position with the value of ``False`` will be unchanged.
+                ..note::
+                    [src/tgt/memory]_mask ensures that position i is allowed to attend the unmasked
```
Comment (janeyx99) [inline torch/nn/modules/transformer.py:174]:
```suggestion
                    [src/tgt/memory]_mask ensures that position i is allowed to attend to the unmasked
```
this grammar feels more right to be but ¯\_(ツ)_/¯

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -158,34 +159,36 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
...
+                    positions. If a BoolTensor is provided, positions with ``True``
+                    are not allowed to attend while ``False`` values will be unchanged. If a FloatTensor
```
Comment (janeyx99) [inline torch/nn/modules/transformer.py:176]: > If a BoolTensor is provided, positions with ``True`` are not allowed to attend while ``False`` values will be unchanged.

What does this mean? How come it feels as if True = not attend AND False is also not attend? Or does "not attend" vary in meaning to "unchanged"?

File: torch/nn/modules/transformer.py (hunk 2 above)
Comment (janeyx99) [reply torch/nn/modules/transformer.py:174]: also, math i?

**Review by janeyx99 — COMMENTED** [review]
looks cool overall, i did get confused by one of the sentences.

File: torch/nn/modules/transformer.py (hunk 1 above)
Comment (mikaylagawarecki) [reply torch/nn/modules/transformer.py:134]: The typing annotations (not here but auto-generated by sphinx) already reflect which args are optional/required

e.g. 
Screenshot 2023-11-29 at 3 47 01 PM

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (mikaylagawarecki) [reply torch/nn/modules/transformer.py]: Good catch, rephrased this

**Pull request closed without merging**